exporter/prometheusremotewriteexporter/wal.go

```diff
@@ -313,113 +322,81 @@ func (prwe *prweWAL) persistToWAL(requests []*prompb.WriteRequest) error {
 			return err
 		}
 		wIndex := prwe.wWALIndex.Add(1)
+		prwe.log.Debug("write", zap.Uint64("index", wIndex))
 		batch.Write(wIndex, protoBlob)
 	}
 
+	// notify possibly waiting tailing routine of write
+	select {
+	case prwe.rNotify <- struct{}{}:
+		// tailing routine is actively waiting for write
+	default:
+		// no receiver, ignore
+	}
+
 	return prwe.wal.WriteBatch(batch)
 }
 
-func (prwe *prweWAL) readPrompbFromWAL(ctx context.Context, index uint64) (wreq *prompb.WriteRequest, err error) {
-	prwe.mu.Lock()
-	defer prwe.mu.Unlock()
-
-	var protoBlob []byte
-	for i := 0; i < 12; i++ {
-		// Firstly check if we've been terminated, then exit if so.
-		select {
-		case <-ctx.Done():
-			return nil, ctx.Err()
-		case <-prwe.stopChan:
-			return nil, fmt.Errorf("attempt to read from WAL after stopped")
-		default:
-		}
-
-		if index <= 0 {
-			index = 1
-		}
-
-		if prwe.wal == nil {
-			return nil, fmt.Errorf("attempt to read from closed WAL")
-		}
+// read repeatedly attempts to fetch a *prompb.WriteRequest from the WAL,
+// aborting on error.
+// prwe.mu is temporarily acquired for the individual read attempts only,
+// allowing writes to happen while waiting
+func (prwe *prweWAL) readPrompbFromWAL(ctx context.Context, index uint64) (*prompb.WriteRequest, error) {
 
-		protoBlob, err = prwe.wal.Read(index)
-		if err == nil { // The read succeeded.
-			req := new(prompb.WriteRequest)
-			if err = proto.Unmarshal(protoBlob, req); err != nil {
-				return nil, err
-			}
+	if prwe == nil {
+		return nil, fmt.Errorf("attempt to read from closed WAL")
+	}
 
-			// Now increment the WAL's read index.
-			prwe.rWALIndex.Add(1)
+	try := func() (*prompb.WriteRequest, error) {
+		prwe.mu.Lock()
+		defer prwe.mu.Unlock()
 
-			return req, nil
+		prwe.log.Debug("read", zap.Uint64("index", index))
+		protoBlob, err := prwe.wal.Read(index)
+		if err != nil {
+			return nil, err
 		}
 
-		if !errors.Is(err, wal.ErrNotFound) {
+		var req prompb.WriteRequest
+		if err := proto.Unmarshal(protoBlob, &req); err != nil {
 			return nil, err
 		}
 
-		if index <= 1 {
-			// This could be the very first attempted read, so try again, after a small sleep.
-			time.Sleep(time.Duration(1<<i) * time.Millisecond)
-			continue
-		}
+		prwe.rWALIndex.Add(1)
+		return &req, nil
+	}
 
-		// Otherwise, we couldn't find the record, let's try watching
-		// the WAL file until perhaps there is a write to it.
-		walWatcher, werr := fsnotify.NewWatcher()
-		if werr != nil {
-			return nil, werr
-		}
-		if werr = walWatcher.Add(prwe.walPath); werr != nil {
-			return nil, werr
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-prwe.stopChan:
+			return nil, fmt.Errorf("attempt to read from WAL after stopped")
```

Comment:
I would suggest using `errors.New` instead of `fmt.Errorf` since it should be "cheaper" to create the value since it doesn't need to parse the string for any formatting symbols.

Alternatively, since this is static you can define the error at the type of the file as part of the "static" vars.